File: shadow/147.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;

        while (head != null) {
            ListNode temp = head.next;
            prev.next = temp;
            insert(head, dummy, temp);
            if (prev.next != temp) {
                prev = prev.next;
            }
            head = temp;
        }

        return dummy.next;
    }

    private void insert(ListNode head, ListNode pointer, ListNode tail) {
        while (pointer.next != tail) {
            if (pointer.next.val < head.val) {
                pointer = pointer.next;
            }
            else {
                ListNode temp = pointer.next;
                pointer.next = head;
                head.next = temp;
                return;
            }
        }
        ListNode temp = pointer.next;
        pointer.next = head;
        head.next = temp;

    }
}

File: Mao/147.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        p = newhead = ListNode(0)
        cur = newhead.next = head
        while cur and cur.next:
            v = cur.next.val
            if v >= cur.val:
                cur = cur.next
                continue
            if p.next.val >= v:
                p = newhead
            while v > p.next.val:
                p = p.next
            temp = cur.next.next
            cur.next.next = p.next
            p.next = cur.next
            cur.next = temp
        return newhead.next
            